### PR TITLE
mantas' packages

### PR DESCRIPTION
**Pull request by mantelllo**
shuffles data at random

### CODE REVIEW

File: Popularity/DCAFPilot/src/python/DCAF/ml/model.py [hunk 1]
```diff
@@ -198,24 +206,67 @@ def model(train_file, newdata_file, idcol, tcol, learner, lparams=None,
...
-        dbses = [int(i) for i in list(tdf[dbs_h])]
-        if  scaler:
-            tdf = getattr(preprocessing, scaler)().fit_transform(tdf)
-        predictions = fit.predict(tdf)
-        data = {'dataset':datasets, dbs_h: dbses, 'prediction':predictions}
-        out = pd.DataFrame(data=data)
-        if  ofile:
-            out.to_csv(ofile, header=True, index=False)
+        nfiles = []
+        if  not ',' in newdata_file:
```
Comment (vkuznet) [inline Popularity/DCAFPilot/src/python/DCAF/ml/model.py]: Can you make it more generic via glob such that input files can be either separated by comma or provided as a pattern, e.g. file*.csv

File: Popularity/DCAFPilot/src/python/DCAF/ml/model.py [hunk 2]
```diff
@@ -198,24 +206,67 @@ def model(train_file, newdata_file, idcol, tcol, learner, lparams=None,
...
+            nfiles = [newdata_file]
+        else:
+            if '(id)' not in ofile:
```
Comment (vkuznet) [inline Popularity/DCAFPilot/src/python/DCAF/ml/model.py]: I'm not sure I understand the logic here, the ofile is an output file, why are you testing for (id) and what (id) means. Please elaborate and make more comments in code.

File: Popularity/DCAFPilot/src/python/DCAF/ml/model.py [hunk 3]
```diff
@@ -198,24 +206,67 @@ def model(train_file, newdata_file, idcol, tcol, learner, lparams=None,
...
+            if  verbose:
+                print(tdf)
+            time0 = time.time()
+            predictions = fit.predict(tdf) if not proba else np.asarray(fit.predict_proba(tdf))[:,list(fit.classes_).index(1)]
+            rtime += time.time()-time0
+            data = {'dataset':datasets, dbs_h: dbses, 'prediction':predictions}
+            out = pd.DataFrame(data=data)
+            if  ofile:
+                out.to_csv(ofile.replace("(clf)", learner).replace("(id)", str(ni)), header=True, index=False)
+            if  timeout: # output running time
```
Comment (vkuznet) [inline Popularity/DCAFPilot/src/python/DCAF/ml/model.py:266]: need explanation what timeout file means and how it should be used

File: Popularity/DCAFPilot/src/python/DCAF/ml/utils.py [hunk 4]
```diff
@@ -136,7 +137,7 @@ def __init__(self, learners=None, scorers=None):
         self.parser.add_option("--train-file", action="store", type="string",
             default="train.csv", dest="train", help="train file, default train.csv")
         self.parser.add_option("--newdata", action="store", type="string",
-            default="", dest="newdata", help="new data file, default None")
+            default="", dest="newdata", help="new data file or provide comma separated file list (then --predict must contain '(id)' to be replaced by file index). Default None")
```
Comment (vkuznet) [inline Popularity/DCAFPilot/src/python/DCAF/ml/utils.py]: show me example how you used (id)

File: Popularity/DCAFPilot/src/python/DCAF/ml/utils.py [hunk 5]
```diff
@@ -157,17 +158,25 @@ def __init__(self, learners=None, scorers=None):
             help="perform grid search, gsearch=<parameters>")
         self.parser.add_option("--predict", action="store", type="string",
             default=None, dest="predict",
-            help="Prediction file name, default None")
+            help="Prediction file name, default None. Use '(model)' for model name or '(id)' for file number if --newdata contains file list")
```
Comment (vkuznet) [inline Popularity/DCAFPilot/src/python/DCAF/ml/utils.py]: need examples of (model) and (id)

File: Popularity/DCAFPilot/src/python/DCAF/ml/model.py [hunk 6]
```diff
@@ -198,24 +207,84 @@ def model(train_file, newdata_file, idcol, tcol, learner, lparams=None,
...
+            dbses = [int(i) for i in list(tdf[dbs_h])]
+            if  scaler:
+                tdf = getattr(preprocessing, scaler)().fit_transform(tdf)
+            if  verbose:
+                print(tdf)
+            time0 = time.time()
+            predictions = fit.predict(tdf) if not proba else np.asarray(fit.predict_proba(tdf))[:,list(fit.classes_).index(1)]
+            rtime += time.time()-time0
+            data = {'dataset':datasets, dbs_h: dbses, 'prediction':predictions}
+            out = pd.DataFrame(data=data)
```
Comment (vkuznet) [inline Popularity/DCAFPilot/src/python/DCAF/ml/model.py]: should be replaced with
out = pd.DataFrame({'dataset':datasets, dbs_h: dbses, 'prediction':predictions})

You created data variable and allocated memory for it, then just copy if to DataFrame. Instead, just create proper dataframe and don't waste program memory allocation for data. You don't even use data variable afterwards in this scope.

File: Popularity/DCAFPilot/src/python/DCAF/ml/model.py [hunk 7]
```diff
@@ -284,7 +353,15 @@ def model_iter(train_file_list, newdata_file, idcol, tcol,
         if  tcol in tdf.columns:
             tdf = tdf.drop(tcol, axis=1)
         datasets = [int(i) for i in list(tdf['dataset'])]
-        dbs_h = 'dbsinst' if 'dbsinst' in tdf.keys() else 'dbs'
+        dfkeys = tdf.keys()
```
Comment (vkuznet) [inline Popularity/DCAFPilot/src/python/DCAF/ml/model.py]: Code repetition, see another comment

File: Popularity/DCAFPilot/src/python/DCAF/ml/model.py [hunk 8]
```diff
@@ -198,24 +207,84 @@ def model(train_file, newdata_file, idcol, tcol, learner, lparams=None,
...
+            tdf = read_data(nfile, drops, scaler=scaler)
+            if  tcol in tdf.columns:
+                tdf = tdf.drop(tcol, axis=1)
+            if  verbose:
+                print("New data file", nfile)
+                print("Columns:", ','.join(tdf.columns))
+                print("test shapes:", tdf.shape)
+            datasets = [int(i) for i in list(tdf['dataset'])]
+            # 20160301 'dbs' header changed to 'dbsinst'
+            dfkeys = tdf.keys()
```
Comment (vkuznet) [inline Popularity/DCAFPilot/src/python/DCAF/ml/model.py]: Code repetition, see another comment. Define function to determine dbs_h header and use it here and in model_iter function below (Line 350).

File: Popularity/DCAFPilot/src/python/DCAF/ml/model.py (hunk 6 above)
Comment (mantelllo) [inline Popularity/DCAFPilot/src/python/DCAF/ml/model.py]: Sure I updated it and will push it with upcoming commit. P.s. This part of code was before me actually, there was just tab added.

Comment (vkuznet) [inline Popularity/DCAFPilot/src/python/DCAF/ml/model.py]: Mantas, you seems to be confused where to place an update. Just check this branch and update code here. Then I can see your changes directly in this PR. It seems to me that this PR will be the first and main one. Then once code is in place you'll tweak other parts. So far I'm making comments before I'll merge your stuff and I want the code to be in a good shape before the merge.

Comment (mantelllo) [inline Popularity/DCAFPilot/src/python/DCAF/ml/model.py]: I merged the requested changes to this branch (commit 5c34b2fa7fda1c878111ed1f59bd6b4e73272f83) and then pull-requested them. Other PR before was made by mistake and it was in turn cancelled. It was done 30 mins before your comment. Did you see the last commit to this branch?

**Comment by vkuznet** [comment]
ok, now PR looks good, merging.
